ttnn/cpp/ttnn/operations/ccl/barrier/device/barrier_op.cpp

```diff
@@ -0,0 +1,116 @@
+        input_tensors.at(0),
+        output_tensors.at(0),
+        this->is_starting_core,
+        this->ring_size,
+        this->ring_index,
+        this->receiver_device_id,
+        this->sender_device_id,
+        this->topology);
+}
+
+namespace operations{
+
+namespace ccl{
+Tensor barrier(
+    const Tensor& input_tensor,
+    const MemoryConfig& output_mem_config,
+    ttnn::ccl::Topology topology)
+{
+    //Host function to launch the OP called from Invoke in barrier.cpp
+
+    //Get the workers
+    auto devices = input_tensor.get_workers();
+
+    //Split the job up between the tensors
+    std::vector<Tensor> output_tensors = {Tensor(operation::get_workers_for_op_output({input_tensor}))};
+
+    //Define the launch_op function
+    operation::launch_op(
+        [output_mem_config, topology, devices](
+            const std::vector<Tensor>& input_tensors,
+            const std::vector<std::optional<const Tensor>>& optional_input_tensors,
+            const std::vector<std::optional<Tensor>>& optional_output_tensors) mutable -> std::vector<Tensor> {
+            bool is_linear = (topology == ttnn::ccl::Topology::Linear);
+
+            const auto& input_tensor = input_tensors.at(0);
+            uint32_t num_devices = devices.size();
+            uint32_t device_index = 0; // Initialize device index
+            std::optional<chip_id_t> receiver_device_id = std::nullopt; // Initialize receiver device ID
+            std::optional<chip_id_t> sender_device_id = std::nullopt; // Initialize sender device ID
+            for (uint32_t i = 0; i < num_devices; ++i) {
+                if (devices.at(i) == input_tensor.device()) {
+                    bool is_last_chip_in_clockwise_direction = is_linear && i == (num_devices - 1);
+                    bool is_last_chip_in_counter_clockwise_direction = is_linear && i == 0;
+                    device_index = i;
+                    receiver_device_id = is_last_chip_in_clockwise_direction ?
+                        std::nullopt :
+                        std::optional<chip_id_t>(devices.at((i + 1) % num_devices)->id());
+                    sender_device_id = is_last_chip_in_counter_clockwise_direction ?
+                        std::nullopt :
+                        std::optional<chip_id_t>(devices.at((i + num_devices - 1) % num_devices)->id());
+                    break;
+                }
+            }
+            return operation::run(
+                ttnn::Barrier{
+                    device_index == 0,
+                    num_devices,
+                    device_index,
+                    receiver_device_id,
+                    sender_device_id,
+                    output_mem_config,
+                    topology},
+                {input_tensor});
+        },
+        {input_tensor},
+        output_tensors);
+    //Return the first output tensor
+    return output_tensors.at(0);
+}
+
+} //namespace ccl end
+
+} //namespace operations end
+} // namespace ttnn end
```

Comment:
Hey @TT-BrianLiu - can you clarify this please because I still see "register_operation" and a launch op call specified there (it's just nested now in a function that's in another file).

Reply:
We need to manually call launch_op as there are slight differences to the launch structure for each of the devices. As much as possible was moved back to the invoke call but there is still a need to modify the default launch_op call. This method was also used in matmul_op.cpp as it also had per-device changes.